### PR TITLE
change intersect order to reduce time

### PR DESCRIPTION
**Pull request by Moelf**
based on #36335

Before:
```
julia> @btime intersect($t, $s);
  42.603 μs (17 allocations: 49.56 KiB)

julia> @btime intersect($s, $t);
  317.603 μs (17 allocations: 49.56 KiB)
```

After:
```
julia> @btime intersect($t, $s);
  53.400 μs (17 allocations: 49.56 KiB)

julia> @btime intersect($s, $t);
  53.687 μs (17 allocations: 49.56 KiB)
```

### CODE REVIEW

**Comment by Moelf** [comment]
a little awkward considering a `intersect(s::AbstractSet, t::AbstractSet)` would be more a specific method thus causing issues.

File: base/abstractset.jl [hunk 1]
```diff
@@ -119,7 +119,7 @@ Set{Int64} with 1 element:
 """
 intersect(s::AbstractSet, itr, itrs...) = intersect!(intersect(s, itr), itrs...)
 intersect(s) = union(s)
-intersect(s::AbstractSet, itr) = mapfilter(_in(s), push!, itr, emptymutable(s))
+intersect(s::AbstractSet, itr) = length(s) > length(itr) ? mapfilter(_in(s), push!, itr, emptymutable(s)) : mapfilter(_in(itr), push!, s, emptymutable(itr))
```
Comment (tkf) [inline base/abstractset.jl:122]: Thanks for the patch!  Unfortunately, this has a few problems:

* `itr` may not define `length`. We need to support generic iterables.  You can use `haslength(itr)` to check if you can call `length(itr)`.
* `emptymutable(itr)` may not return a set. We should always use `emptymutable(s)`.

Comment (tkf) [reply base/abstractset.jl:122]: Also, we need to check `hasfastin(itr)` if we can use `_in(itr)`.

Finally, please limit the length of a line to 92 characters (as mentioned in CONTRIBUTING.md)

Comment (Moelf) [reply base/abstractset.jl]: does it look better now? (force pushed

File: base/abstractset.jl [hunk 2]
```diff
@@ -119,7 +119,15 @@ Set{Int64} with 1 element:
...
 intersect(s) = union(s)
-intersect(s::AbstractSet, itr) = mapfilter(_in(s), push!, itr, emptymutable(s))
+function intersect(s::AbstractSet, itr) 
+    if haslength(itr) && hasfastin(itr)
+        if length(s) > length(itr) 
+            return mapfilter(_in(s), push!, itr, emptymutable(s)) 
+        else
+            return mapfilter(_in(itr), push!, s, emptymutable(s))
+        end
+    end
```
Comment (tkf) [inline base/abstractset.jl:129]: You need something like

```suggestion
    if hasfastin(itr) && haslength(itr) && length(s) < length(itr) 
        return mapfilter(_in(itr), push!, s, emptymutable(s))
    else
        return mapfilter(_in(s), push!, itr, emptymutable(s)) 
    end
```

Comment (tkf) [reply base/abstractset.jl:129]: Also please make sure both branches are tested.

**Comment by CameronBieganek** [comment]
By the way, it looks like the `_in` function defined on line 46 of abstractset.jl is just the curried version of `in`, which now exists. So it seems like `_in(itr)` could be replaced with `in(itr)`, unless I'm missing something...

Also, it appears that `intersect!` suffers from the same problem as `intersect`:

```julia
julia> @benchmark intersect!(s, t) setup=(s = Set(rand(1:10_000, 1000))) setup=(t = Set(1000:9000)) evals=1
BenchmarkTools.Trial: 
  memory estimate:  32 bytes
  allocs estimate:  2
  --------------
  minimum time:     11.558 μs (0.00% GC)
  median time:      12.407 μs (0.00% GC)
  mean time:        13.226 μs (0.00% GC)
  maximum time:     80.413 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark intersect!(t, s) setup=(s = Set(rand(1:10_000, 1000))) setup=(t = Set(1000:9000)) evals=1
BenchmarkTools.Trial: 
  memory estimate:  32 bytes
  allocs estimate:  2
  --------------
  minimum time:     219.543 μs (0.00% GC)
  median time:      222.558 μs (0.00% GC)
  mean time:        230.665 μs (1.53% GC)
  maximum time:     3.513 ms (93.35% GC)
  --------------
  samples:          9632
  evals/sample:     1
```

So it would be nice if we could also fix that in this PR. :)

**Comment by tkf** [comment]
You can't flip the order of `intersect!` because the first argument has to be mutated.  Not very sure if we can do something with `intersect!(s, t)`.  It might be possible to get some speedup from `intersect!(a, b, c)` by flipping `b` and `c` though.

File: base/abstractset.jl (hunk 2 above)
Comment (Moelf) [reply base/abstractset.jl]: sorry for the delay, it should be fine now also added tests for it

**Comment by Moelf** [comment]
im not sure how to deal with `intersect(s::AbstractSet, itr, itrs...)`, the problem is that `itrs` is a tuple and to swap place of `itr` and the shortest element in `itrs` seems expensive?

**Comment by Moelf** [comment]
ok I tried to add short cut for `intersect(a,b,c...)` as well plz review

```
julia> s1 = Set(rand(1:100_000, 10000));

julia> s2 = Set(rand(1:100_000, 30000));

julia> s3 = Set(rand(1:100_000, 50000));

julia> t = Set(10000:90000);

julia> @btime intersect($t, $s1, $s2, $s3);
  1.199 ms (29 allocations: 193.98 KiB)

julia> @btime intersect($t, $s3, $s2, $s1);
  4.596 ms (34 allocations: 770.28 KiB)
```

after:
```
julia> @btime intersect($t, $s1, $s2, $s3);
  1.112 ms (22 allocations: 193.86 KiB)

julia> @btime intersect($t, $s3, $s2, $s1);
  1.112 ms (31 allocations: 194.02 KiB)
```

**Comment by laborg** [comment]
This might make `intersect` slower for small sets:
1.4.2:
```
julia> a = [Set([7, 9, 8]),
           Set([7, 4, 2, 8, 6]),
           Set([9, 10, 2, 5, 8, 6, 1]),
           Set([7, 4, 9, 2, 8, 1]),
           Set([7, 9, 4, 10, 2, 8, 5, 6]),
           Set([7, 4, 10, 2, 8, 6, 1]),
           Set([9, 4, 3, 5, 8, 6, 1]),
           Set([4, 2, 5, 7, 9, 10, 8, 1]),
           Set([4, 2, 3, 7, 9, 10, 8, 6, 1])];

julia> b = reverse(a);

julia> @btime intersect($a...)
  750.129 ns (22 allocations: 816 bytes)
Set{Int64} with 1 element:
  8

julia> @btime intersect($b...)
  865.911 ns (22 allocations: 816 bytes)
Set{Int64} with 1 element:
  8
```
This PR:
```
julia> @btime intersect($a...)
  1.216 μs (27 allocations: 944 bytes)
Set{Int64} with 1 element:
  8

julia> @btime intersect($b...)

  2.220 μs (33 allocations: 1.25 KiB)
Set{Int64} with 1 element:
  8
```

**Comment by Moelf** [comment]
yea this is inevitable, what can be done is add a ` length(s) > 100 &&`; the thing is as soon as the the leading `Set` has a small size, the process would be very fast, so only really need to swap `itr, itrs` if `s` is long .

**Comment by Moelf** [comment]
after rebase the tests are all passed now

**Comment by CameronBieganek** [comment]
Just curious, what's the status of this PR?

**Comment by StefanKarpinski** [comment]
Appears to be in good shape but needs review.

**Comment by oscardssmith** [comment]
@StefanKarpinski why does this need the triage label?

**Comment by StefanKarpinski** [comment]
Seemed like a good way to get some collective code review on it.

File: base/abstractset.jl [hunk 3]
```diff
@@ -117,9 +117,28 @@ Set{Int64} with 1 element:
...
-intersect(s::AbstractSet, itr, itrs...) = intersect!(intersect(s, itr), itrs...)
+function intersect(s::AbstractSet, itr, itrs...)
+    # determine if swap order is useful and viable
+    if length(s)>50 && haslength(itr) && all(haslength.(itrs))
+        itrs_lengths = length.(itrs)
+        min_length = minimum(itrs_lengths)
+        # do nothing if s is longer than the rest or itr is already the shortest
+        if length(itr) > min_length >= length(s)
+            min_idx = argmin(itrs_lengths)
+            new_itrs = (itrs[1:min_idx-1]..., itr, itrs[min_idx+1:end]...)
```
Comment (vtjnash) [inline base/abstractset.jl:128]:
```suggestion
            new_itrs = setindex(itrs, itr, min_idx)
```

File: base/abstractset.jl [hunk 4]
```diff
@@ -117,9 +117,28 @@ Set{Int64} with 1 element:
   2
 ```
 """
-intersect(s::AbstractSet, itr, itrs...) = intersect!(intersect(s, itr), itrs...)
+function intersect(s::AbstractSet, itr, itrs...)
+    # determine if swap order is useful and viable
+    if length(s)>50 && haslength(itr) && all(haslength.(itrs))
```
Comment (vtjnash) [inline base/abstractset.jl:122]:
```suggestion
    if length(s)>50 && haslength(itr) && all(map(haslength, itrs))
```

File: base/abstractset.jl [hunk 5]
```diff
@@ -117,9 +117,28 @@ Set{Int64} with 1 element:
...
+        itrs_lengths = length.(itrs)
```
Comment (vtjnash) [inline base/abstractset.jl:123]:
```suggestion
        itrs_lengths = map(length, itrs)
```

File: base/abstractset.jl [hunk 6]
```diff
@@ -117,9 +117,28 @@ Set{Int64} with 1 element:
...
+        min_length = minimum(itrs_lengths)
```
Comment (vtjnash) [inline base/abstractset.jl:124]:
```suggestion
        min_idx = argmin(itrs_lengths)
        min_length = itrs_lengths[min_idx]
```

File: base/abstractset.jl [hunk 7]
```diff
@@ -117,9 +117,28 @@ Set{Int64} with 1 element:
...
+        # do nothing if s is longer than the rest or itr is already the shortest
+        if length(itr) > min_length >= length(s)
+            min_idx = argmin(itrs_lengths)
```
Comment (vtjnash) [inline base/abstractset.jl:127]:
```suggestion
```

**Comment by vtjnash** [comment]
I can't see a reason that triage is necessary either. I've made a few small comments. If you could test that and rebase, we can merge soon.

File: base/abstractset.jl (hunk 3 above)
Comment (Moelf) [reply base/abstractset.jl:128]: `setindex!`？

**Comment by oscardssmith** [comment]
Is this good to merge?

File: base/abstractset.jl [hunk 8]
```diff
@@ -117,9 +117,28 @@ Set{Int64} with 1 element:
   2
 ```
 """
-intersect(s::AbstractSet, itr, itrs...) = intersect!(intersect(s, itr), itrs...)
+function intersect(s::AbstractSet, itr, itrs...)
+    # determine if swap order is useful and viable
+    if length(s)>50 && haslength(itr) && all(map(haslength, itrs))
```
Comment (JeffBezanson) [inline base/abstractset.jl:122]:
```suggestion
    if length(s)>50 && haslength(itr) && all(haslength, itrs)
```

File: base/abstractset.jl [hunk 9]
```diff
@@ -117,9 +117,28 @@ Set{Int64} with 1 element:
...
+        itrs_lengths = map(length, itrs)
+        min_idx = argmin(itrs_lengths)
+        min_length = itrs_lengths[min_idx]
+        # do nothing if s is longer than the rest or itr is already the shortest
+        if length(itr) > min_length >= length(s)
```
Comment (JeffBezanson) [inline base/abstractset.jl:127]: Why is this looking at `length(s)`? Why not always intersect `s` with the shortest itr?

**Comment by JeffBezanson** [comment]
Needs a small change & rebase, then we can merge.

**Comment by Moelf** [comment]
oops, my hacky edit has resulted in weird diff but it should be fine with respect to master branch and both test cases (two-element `interset` and multi-element `intersect(s, itr, itrs...)` work now):
```

julia> @btime intersect($t, $s);
  14.217 μs (17 allocations: 49.14 KiB)

julia> @btime intersect($s, $t);
  14.828 μs (17 allocations: 49.14 KiB)

julia> @btime intersect($t, $s1, $s2, $s3);
  534.275 μs (22 allocations: 193.34 KiB)

julia> @btime intersect($t, $s3, $s2, $s1);
  520.268 μs (22 allocations: 193.34 KiB)
```

**Pull request merged**